Apply mypy in pre-commit/CI

Typehints are already good in this codebase but it would be useful to add in [mypy](https://www.mypy-lang.org/) checks
via a pre-commit hook.

``` yaml
  - repo: https://github.com/pre-commit/mirrors-mypy
    rev: v1.14.1
    hooks:
      - id: mypy
        args: [--ignore-missing-imports] # default but useful to be explicit
```

To help with this we can try running [RighTyper](https://github.com/RightTyper/RightTyper) against the code base as it
analyses and informs what the typehints should be for a code base. It only works with Python >= 3.12 though (and
currently we can not support Python >= 3.12 with TopoStats due to as yet unresolved issues with the `topoly` dependency,
see [Performance regression of Topoly with Python 3.12 · Issue #927 ·
AFM-SPM/TopoStats](https://github.com/AFM-SPM/TopoStats/issues/927)). However we should be able to add support for
Python 3.12 in AFMReader.

- [x] #97

First run of `mypy` against the code base reveals the following errors...

``` bash
❱ pre-commit run mypy --all-files
mypy.....................................................................Failed
- hook id: mypy
- exit code: 1

tests/test_jpk.py:35: error: Incompatible types in assignment (expression has type "float", variable has type "type[float]")  [assignment]
tests/test_ibw.py:31: error: Incompatible types in assignment (expression has type "float", variable has type "type[float]")  [assignment]
AFMReader/io.py:287: error: Unsupported operand types for + ("bytes" and "str")  [operator]
AFMReader/io.py:318: error: "str" has no attribute "decode"; maybe "encode"?  [attr-defined]
tests/test_io.py:149: error: Value of type "object" is not indexable  [index]
tests/test_io.py:150: error: Value of type "object" is not indexable  [index]
tests/test_io.py:180: error: Value of type "object" is not indexable  [index]
tests/test_io.py:181: error: Value of type "object" is not indexable  [index]
AFMReader/gwy.py:48: error: Item "str" of "Path | str" has no attribute "stem"  [union-attr]
AFMReader/gwy.py:50: error: Need type annotation for "image_data_dict" (hint: "image_data_dict: dict[<type>, <type>] = ...")  [var-annotated]
AFMReader/gwy.py:51: error: Argument 1 to "open" of "Path" has incompatible type "Path | str"; expected "Path"  [arg-type]
AFMReader/gwy.py:56: error: Argument 1 to "gwy_read_object" has incompatible type "IO[Any]"; expected "TextIOWrapper[_WrappedBuffer]"  [arg-type]
AFMReader/gwy.py:66: error: Item "str" of "Path | str" has no attribute "suffix"  [union-attr]
AFMReader/gwy.py:104: error: Argument 1 to "read_uint32" has incompatible type "TextIOWrapper[_WrappedBuffer]"; expected "BinaryIO"  [arg-type]
AFMReader/gwy.py:140: error: Need type annotation for "sub_dict" (hint: "sub_dict: dict[<type>, <type>] = ...")  [var-annotated]
AFMReader/gwy.py:148: error: Incompatible types in assignment (expression has type "int", variable has type "str")  [assignment]
AFMReader/gwy.py:148: error: Argument "open_file" to "read_uint32" has incompatible type "TextIOWrapper[_WrappedBuffer]"; expected "BinaryIO"  [arg-type]
AFMReader/gwy.py:152: error: Incompatible types in assignment (expression has type "float", variable has type "str")  [assignment]
AFMReader/gwy.py:152: error: Argument "open_file" to "read_double" has incompatible type "TextIOWrapper[_WrappedBuffer]"; expected "BinaryIO"  [arg-type]
AFMReader/gwy.py:160: error: Argument "open_file" to "read_uint32" has incompatible type "TextIOWrapper[_WrappedBuffer]"; expected "BinaryIO"  [arg-type]
AFMReader/gwy.py:165: error: Argument "open_file" to "read_double" has incompatible type "TextIOWrapper[_WrappedBuffer]"; expected "BinaryIO"  [arg-type]
AFMReader/gwy.py:239: error: "str" has no attribute "decode"; maybe "encode"?  [attr-defined]
AFMReader/asd.py:324: error: Incompatible types in assignment (expression has type "int", target has type "str")  [assignment]
AFMReader/asd.py:327: error: Incompatible types in assignment (expression has type "int", target has type "str")  [assignment]
AFMReader/asd.py:329: error: Incompatible types in assignment (expression has type "int", target has type "str")  [assignment]
AFMReader/asd.py:330: error: Incompatible types in assignment (expression has type "int", target has type "str")  [assignment]
AFMReader/asd.py:332: error: Incompatible types in assignment (expression has type "int", target has type "str")  [assignment]
AFMReader/asd.py:334: error: Incompatible types in assignment (expression has type "int", target has type "str")  [assignment]
AFMReader/asd.py:335: error: Incompatible types in assignment (expression has type "int", target has type "str")  [assignment]
AFMReader/asd.py:337: error: Incompatible types in assignment (expression has type "int", target has type "str")  [assignment]
AFMReader/asd.py:338: error: Incompatible types in assignment (expression has type "int", target has type "str")  [assignment]
AFMReader/asd.py:340: error: Incompatible types in assignment (expression has type "float", target has type "str")  [assignment]
AFMReader/asd.py:342: error: Incompatible types in assignment (expression has type "float", target has type "str")  [assignment]
AFMReader/asd.py:344: error: Incompatible types in assignment (expression has type "float", target has type "str")  [assignment]
AFMReader/asd.py:349: error: Incompatible types in assignment (expression has type "int", target has type "str")  [assignment]
AFMReader/asd.py:350: error: Incompatible types in assignment (expression has type "int", target has type "str")  [assignment]
AFMReader/asd.py:350: error: Unsupported operand types for ^ ("int" and "str")  [operator]
AFMReader/asd.py:352: error: Incompatible types in assignment (expression has type "bool", target has type "str")  [assignment]
AFMReader/asd.py:354: error: Incompatible types in assignment (expression has type "int", target has type "str")  [assignment]
AFMReader/asd.py:358: error: Incompatible types in assignment (expression has type "int", target has type "str")  [assignment]
AFMReader/asd.py:359: error: Incompatible types in assignment (expression has type "int", target has type "str")  [assignment]
AFMReader/asd.py:360: error: Incompatible types in assignment (expression has type "int", target has type "str")  [assignment]
AFMReader/asd.py:361: error: Incompatible types in assignment (expression has type "int", target has type "str")  [assignment]
AFMReader/asd.py:362: error: Incompatible types in assignment (expression has type "int", target has type "str")  [assignment]
AFMReader/asd.py:363: error: Incompatible types in assignment (expression has type "int", target has type "str")  [assignment]
AFMReader/asd.py:365: error: Incompatible types in assignment (expression has type "int", target has type "str")  [assignment]
AFMReader/asd.py:367: error: Incompatible types in assignment (expression has type "float", target has type "str")  [assignment]
AFMReader/asd.py:368: error: Incompatible types in assignment (expression has type "float", target has type "str")  [assignment]
AFMReader/asd.py:374: error: Incompatible types in assignment (expression has type "int", target has type "str")  [assignment]
AFMReader/asd.py:376: error: Incompatible types in assignment (expression has type "int", target has type "str")  [assignment]
AFMReader/asd.py:378: error: Incompatible types in assignment (expression has type "int", target has type "str")  [assignment]
AFMReader/asd.py:380: error: Incompatible types in assignment (expression has type "int", target has type "str")  [assignment]
AFMReader/asd.py:382: error: Argument "length_bytes" to "read_null_separated_utf8" has incompatible type "str"; expected "int"  [arg-type]
AFMReader/asd.py:384: error: Incompatible types in assignment (expression has type "float", target has type "str")  [assignment]
AFMReader/asd.py:386: error: Incompatible types in assignment (expression has type "float", target has type "str")  [assignment]
AFMReader/asd.py:388: error: Incompatible types in assignment (expression has type "int", target has type "str")  [assignment]
AFMReader/asd.py:390: error: Argument 2 to "skip_bytes" has incompatible type "str"; expected "int"  [arg-type]
AFMReader/asd.py:393: error: No overload variant of "range" matches argument type "str"  [call-overload]
AFMReader/asd.py:393: note: Possible overload variants:
AFMReader/asd.py:393: note:     def range(SupportsIndex, /) -> range
AFMReader/asd.py:393: note:     def range(SupportsIndex, SupportsIndex, SupportsIndex = ..., /) -> range
AFMReader/asd.py:433: error: Incompatible types in assignment (expression has type "str", target has type "int")  [assignment]
AFMReader/asd.py:434: error: Incompatible types in assignment (expression has type "str", target has type "int")  [assignment]
AFMReader/asd.py:464: error: Incompatible types in assignment (expression has type "float", target has type "int")  [assignment]
AFMReader/asd.py:466: error: Incompatible types in assignment (expression has type "float", target has type "int")  [assignment]
AFMReader/asd.py:468: error: Incompatible types in assignment (expression has type "float", target has type "int")  [assignment]
AFMReader/asd.py:476: error: Incompatible types in assignment (expression has type "str", target has type "int")  [assignment]
AFMReader/asd.py:482: error: Incompatible types in assignment (expression has type "float", target has type "int")  [assignment]
AFMReader/asd.py:483: error: Incompatible types in assignment (expression has type "float", target has type "int")  [assignment]
AFMReader/asd.py:485: error: Incompatible types in assignment (expression has type "float", target has type "int")  [assignment]
AFMReader/asd.py:486: error: Incompatible types in assignment (expression has type "float", target has type "int")  [assignment]
AFMReader/asd.py:487: error: Incompatible types in assignment (expression has type "float", target has type "int")  [assignment]
AFMReader/asd.py:489: error: Incompatible types in assignment (expression has type "float", target has type "int")  [assignment]
AFMReader/asd.py:495: error: Incompatible types in assignment (expression has type "str", target has type "int")  [assignment]
AFMReader/asd.py:501: error: Incompatible types in assignment (expression has type "str", target has type "int")  [assignment]
AFMReader/asd.py:539: error: Incompatible types in assignment (expression has type "str", target has type "int")  [assignment]
AFMReader/asd.py:540: error: Incompatible types in assignment (expression has type "str", target has type "int")  [assignment]
AFMReader/asd.py:570: error: Incompatible types in assignment (expression has type "float", target has type "int")  [assignment]
AFMReader/asd.py:572: error: Incompatible types in assignment (expression has type "float", target has type "int")  [assignment]
AFMReader/asd.py:574: error: Incompatible types in assignment (expression has type "float", target has type "int")  [assignment]
AFMReader/asd.py:582: error: Incompatible types in assignment (expression has type "str", target has type "int")  [assignment]
AFMReader/asd.py:588: error: Incompatible types in assignment (expression has type "float", target has type "int")  [assignment]
AFMReader/asd.py:589: error: Incompatible types in assignment (expression has type "float", target has type "int")  [assignment]
AFMReader/asd.py:591: error: Incompatible types in assignment (expression has type "float", target has type "int")  [assignment]
AFMReader/asd.py:592: error: Incompatible types in assignment (expression has type "float", target has type "int")  [assignment]
AFMReader/asd.py:593: error: Incompatible types in assignment (expression has type "float", target has type "int")  [assignment]
AFMReader/asd.py:595: error: Incompatible types in assignment (expression has type "float", target has type "int")  [assignment]
AFMReader/asd.py:601: error: Incompatible types in assignment (expression has type "str", target has type "int")  [assignment]
AFMReader/asd.py:607: error: Incompatible types in assignment (expression has type "str", target has type "int")  [assignment]
AFMReader/asd.py:614: error: Incompatible types in assignment (expression has type "float", target has type "int")  [assignment]
AFMReader/asd.py:686: error: Incompatible types in assignment (expression has type "float", target has type "int")  [assignment]
AFMReader/asd.py:687: error: Incompatible types in assignment (expression has type "float", target has type "int")  [assignment]
AFMReader/asd.py:702: error: "VoltageLevelConverter" has no attribute "level_to_voltage"  [attr-defined]
AFMReader/asd.py:704: error: "bytes" has no attribute "reshape"  [attr-defined]
tests/test_topostats.py:47: error: Missing positional argument "self" in call to "keys" of "dict"  [call-arg]
tests/test_gwy.py:31: error: Need type annotation for "test_dict" (hint: "test_dict: dict[<type>, <type>] = ...")  [var-annotated]
tests/test_gwy.py:32: error: Argument "open_file" to "gwy_read_object" has incompatible type "BufferedReader"; expected "TextIOWrapper[_WrappedBuffer]"  [arg-type]
tests/test_gwy.py:42: error: Need type annotation for "test_dict" (hint: "test_dict: dict[<type>, <type>] = ...")  [var-annotated]
tests/test_gwy.py:43: error: Argument "open_file" to "gwy_read_component" has incompatible type "BufferedReader"; expected "TextIOWrapper[_WrappedBuffer]"  [arg-type]
tests/test_gwy.py:111: error: Argument 1 to "read_gwy_component_dtype" has incompatible type "BufferedReader"; expected "TextIOWrapper[_WrappedBuffer]"  [arg-type]
tests/test_asd.py:28: error: Argument 1 to "len" has incompatible type "type[list[Any]]"; expected "Sized"  [arg-type]
Found 98 errors in 9 files (checked 17 source files)
```

--- a/.pre-commit-config.yaml
+++ b/.pre-commit-config.yaml
@@ -56,6 +56,12 @@ repos:
     hooks:
       - id: numpydoc-validation
 
+  - repo: https://github.com/pre-commit/mirrors-mypy
+    rev: v1.14.1
+    hooks:
+      - id: mypy
+        args: [--ignore-missing-imports] # default but useful to be explicit
+
   - repo: local
     hooks:
       - id: pylint

--- a/.pylintrc
+++ b/.pylintrc
@@ -145,11 +145,11 @@ never-returning-functions=sys.exit,argparse.parse_error
 
 # The type of string formatting that logging methods do. `old` means using %
 # formatting, `new` is for `{}` formatting.
-logging-format-style=old
+logging-format-style=new
 
 # Logging modules to check that the string format arguments are in logging
 # function parameter format.
-logging-modules=logging
+logging-modules=loguru
 
 
 [SPELLING]
@@ -573,6 +573,8 @@ max-statements=50
 # Minimum number of public methods for a class (see R0903).
 min-public-methods=2
 
+# Maximum positional arguments
+max-positional-arguments=6
 
 
 [EXCEPTIONS]

--- a/AFMReader/asd.py
+++ b/AFMReader/asd.py
@@ -3,7 +3,7 @@
 from __future__ import annotations
 from pathlib import Path
 
-from typing import BinaryIO
+from typing import BinaryIO, Self
 
 
 import numpy as np
@@ -50,7 +50,7 @@ class VoltageLevelConverter:
         values. Typically 12, hence 2^12 = 4096 sensitivity levels.
     """
 
-    def __init__(self, analogue_digital_range: float, scaling_factor: float, resolution: int) -> None:
+    def __init__(self: BipolarConverter, analogue_digital_range: float, scaling_factor: float, resolution: int) -> None:
         """
         Convert arbitrary height levels from the AFM into real world nanometre heights.
 
@@ -78,7 +78,7 @@ class VoltageLevelConverter:
 class UnipolarConverter(VoltageLevelConverter):
     """A VoltageLevelConverter for unipolar encodings. (0 to +X Volts)."""
 
-    def level_to_voltage(self, level: float) -> float:
+    def level_to_voltage(self: Self, level: float) -> float:
         """
         Calculate the real world height scale in nanometres for an arbitrary level value.
 
@@ -101,7 +101,7 @@ class UnipolarConverter(VoltageLevelConverter):
 class BipolarConverter(VoltageLevelConverter):
     """A VoltageLevelConverter for bipolar encodings. (-X to +X Volts)."""
 
-    def level_to_voltage(self, level: float) -> float:
+    def level_to_voltage(self: Self, level: float) -> float:
         """
         Calculate the real world height scale in nanometres for an arbitrary level value.
 
@@ -398,7 +398,7 @@ def read_header_file_version_0(open_file: BinaryIO) -> dict:
 
 
 # pylint: disable=too-many-statements
-def read_header_file_version_1(open_file: BinaryIO):
+def read_header_file_version_1(open_file: BinaryIO) -> dict[str, int]:
     """
     Read the header metadata for a .asd file using file version 1.
 

--- a/AFMReader/asd.py
+++ b/AFMReader/asd.py
@@ -2,8 +2,13 @@
 
 from __future__ import annotations
 from pathlib import Path
+import sys
 
-from typing import BinaryIO, Self
+if sys.version_info.minor < 11:
+    from typing import Any, BinaryIO
+    from typing_extensions import Self
+else:
+    from typing import Any, BinaryIO, Self
 
 
 import numpy as np
@@ -29,6 +34,8 @@ from AFMReader.io import (
 
 logger.enable(__package__)
 
+# mypy: disable-error-code="assignment"
+
 
 # pylint: disable=too-few-public-methods
 class VoltageLevelConverter:
@@ -50,7 +57,7 @@ class VoltageLevelConverter:
         values. Typically 12, hence 2^12 = 4096 sensitivity levels.
     """
 
-    def __init__(self: BipolarConverter, analogue_digital_range: float, scaling_factor: float, resolution: int) -> None:
+    def __init__(self: Self, analogue_digital_range: float, scaling_factor: float, resolution: int) -> None:
         """
         Convert arbitrary height levels from the AFM into real world nanometre heights.
 
@@ -311,7 +318,7 @@ def read_header_file_version_0(open_file: BinaryIO) -> dict:
     dict
         Dictionary of metadata decoded from the file header.
     """
-    header_dict = {}
+    header_dict: dict[str, Any] = {}
 
     # There only ever seem to be two channels available
     # Channel encoding are all in LITTLE ENDIAN format.
@@ -699,9 +706,9 @@ def read_channel_data(
         # Decode frame data from bytes. Data is always stored in signed 2 byte integer form
         frame_data = np.frombuffer(frame_data, dtype=np.int16)
         # Convert from Voltage to Real units
-        frame_data = analogue_digital_converter.level_to_voltage(frame_data)
+        frame_data = analogue_digital_converter.level_to_voltage(frame_data)  # type: ignore
         # Reshape frame to 2D array
-        frame_data = frame_data.reshape((y_pixels, x_pixels))
+        frame_data = frame_data.reshape((y_pixels, x_pixels))  # type: ignore
         frames.append(frame_data)
 
     return frames

--- a/AFMReader/gwy.py
+++ b/AFMReader/gwy.py
@@ -1,10 +1,9 @@
 """For decoding and loading .gwy AFM file format into Python Numpy arrays."""
 
 from __future__ import annotations
-import io
 from pathlib import Path
 import re
-
+from typing import Any, BinaryIO
 
 from loguru import logger
 import numpy as np
@@ -12,7 +11,7 @@ import numpy as np
 from AFMReader.io import read_uint32, read_null_terminated_string, read_char, read_double
 
 
-def load_gwy(file_path: Path | str, channel: str):
+def load_gwy(file_path: Path | str, channel: str) -> tuple[np.ndarray[Any, np.dtypes.Float64DType], float]:
     """
     Extract image and pixel to nm scaling from the .gwy file.
 
@@ -45,13 +44,14 @@ def load_gwy(file_path: Path | str, channel: str):
     ```
     """
     logger.info(f"Loading image from : {file_path}")
+    file_path = Path(file_path)
     filename = file_path.stem
     try:
-        image_data_dict = {}
+        image_data_dict: dict[Any, Any] = {}
         with Path.open(file_path, "rb") as open_file:  # pylint: disable=unspecified-encoding
             # Read header
             header = open_file.read(4)
-            logger.debug(f"Gwy file header: {header}")
+            logger.debug(f"Gwy file header: {header.decode}")
 
             gwy_read_object(open_file, data_dict=image_data_dict)
 
@@ -89,13 +89,13 @@ def load_gwy(file_path: Path | str, channel: str):
     return (image, px_to_nm)
 
 
-def gwy_read_object(open_file: io.TextIOWrapper, data_dict: dict) -> None:
+def gwy_read_object(open_file: BinaryIO, data_dict: dict) -> None:
     """
     Parse and extract data from a `.gwy` file object, starting at the current open file read position.
 
     Parameters
     ----------
-    open_file : io.TextIOWrapper
+    open_file : BinaryIO
         An open file object.
     data_dict : dict
         Dictionary of `.gwy` file image properties.
@@ -114,13 +114,13 @@ def gwy_read_object(open_file: io.TextIOWrapper, data_dict: dict) -> None:
         read_data_size += component_data_size
 
 
-def gwy_read_component(open_file: io.TextIOWrapper, initial_byte_pos: int, data_dict: dict) -> int:
+def gwy_read_component(open_file: BinaryIO, initial_byte_pos: int, data_dict: dict) -> int:
     """
     Parse and extract data from a `.gwy` file object, starting at the current open file read position.
 
     Parameters
     ----------
-    open_file : io.TextIOWrapper
+    open_file : BinaryIO
         An open file object.
     initial_byte_pos : int
         Initial position, as byte.
@@ -137,7 +137,7 @@ def gwy_read_component(open_file: io.TextIOWrapper, initial_byte_pos: int, data_
 
     if data_type == "o":
         logger.debug(f"component name: {component_name} | dtype: {data_type} |")
-        sub_dict = {}
+        sub_dict: dict[Any, Any] = {}
         gwy_read_object(open_file=open_file, data_dict=sub_dict)
         data_dict[component_name] = sub_dict
     elif data_type == "c":
@@ -145,11 +145,11 @@ def gwy_read_component(open_file: io.TextIOWrapper, initial_byte_pos: int, data_
         logger.debug(f"component name: {component_name} | dtype: {data_type} | value: {value}")
         data_dict[component_name] = value
     elif data_type == "i":
-        value = read_uint32(open_file=open_file)
+        value = read_uint32(open_file=open_file)  # type: ignore
         logger.debug(f"component name: {component_name} | dtype: {data_type} | value: {value}")
         data_dict[component_name] = value
     elif data_type == "d":
-        value = read_double(open_file=open_file)
+        value = read_double(open_file=open_file)  # type: ignore
         logger.debug(f"component name: {component_name} | dtype: {data_type} | value: {value}")
         data_dict[component_name] = value
     elif data_type == "s":
@@ -202,7 +202,7 @@ def gwy_get_channels(gwy_file_structure: dict) -> dict:
     return channel_ids
 
 
-def read_gwy_component_dtype(open_file: io.TextIOWrapper) -> str:
+def read_gwy_component_dtype(open_file: BinaryIO) -> str:
     """
     Read the data type of a `.gwy` file component.
 
@@ -228,7 +228,7 @@ def read_gwy_component_dtype(open_file: io.TextIOWrapper) -> str:
 
     Parameters
     ----------
-    open_file : io.TextIOWrapper
+    open_file : BinaryIO
         An open file object.
 
     Returns

--- a/AFMReader/io.py
+++ b/AFMReader/io.py
@@ -1,6 +1,5 @@
 """For reading and writing data from / to files."""
 
-import io
 import struct
 from typing import BinaryIO
 import h5py
@@ -260,13 +259,13 @@ def unpack_hdf5(open_hdf5_file: h5py.File, group_path: str = "/") -> dict:
     return data
 
 
-def read_null_terminated_string(open_file: io.TextIOWrapper, encoding: str = "utf-8") -> str:
+def read_null_terminated_string(open_file: BinaryIO, encoding: str = "utf-8") -> str:
     """
     Read an open file from the current position in the open binary file, until the next null value.
 
     Parameters
     ----------
-    open_file : io.TextIOWrapper
+    open_file : BinaryIO
         An open file object.
     encoding : str
         Encoding to use when decoding the bytes.
@@ -301,13 +300,13 @@ def read_null_terminated_string(open_file: io.TextIOWrapper, encoding: str = "ut
         raise e
 
 
-def read_char(open_file: io.TextIOWrapper) -> str:
+def read_char(open_file: BinaryIO) -> str:
     """
     Read a character from an open binary file.
 
     Parameters
     ----------
-    open_file : io.TextIOWrapper
+    open_file : BinaryIO
         An open file object.
 
     Returns

--- a/tests/test_asd.py
+++ b/tests/test_asd.py
@@ -25,6 +25,6 @@ def test_load_asd(file_name: str, channel: str, number_of_frames: int, pixel_to_
     file_path = RESOURCES / file_name
     result_frames, result_pixel_to_nm_scaling, result_metadata = load_asd(file_path, channel)
 
-    assert len(result_frames) == number_of_frames
+    assert len(result_frames) == number_of_frames  # type: ignore
     assert result_pixel_to_nm_scaling == pixel_to_nm_scaling
     assert isinstance(result_metadata, dict)

--- a/tests/test_gwy.py
+++ b/tests/test_gwy.py
@@ -1,7 +1,7 @@
 """Test the loading of gwy files."""
 
 from pathlib import Path
-
+from typing import Any
 
 import pytest
 import numpy as np
@@ -12,7 +12,7 @@ BASE_DIR = Path.cwd()
 RESOURCES = BASE_DIR / "tests" / "resources"
 
 
-def test_load_gwy():
+def test_load_gwy() -> None:
     """Test the normal operation of loading a .gwy file."""
     channel = "ZSensor"
     file_path = RESOURCES / "sample_0.gwy"
@@ -28,7 +28,7 @@ def test_gwy_read_object() -> None:
     """Test reading an object of a `.gwy` file object from an open binary file."""
     with Path.open(RESOURCES / "IO_binary_file.bin", "rb") as open_binary_file:  # pylint: disable=unspecified-encoding
         open_binary_file.seek(19)
-        test_dict = {}
+        test_dict: dict[Any, Any] = {}
         gwy.gwy_read_object(open_file=open_binary_file, data_dict=test_dict)
 
         assert list(test_dict.keys()) == ["test component", "test object component"]
@@ -39,7 +39,7 @@ def test_gwy_read_component() -> None:
     """Tests reading a component of a `.gwy` file object from an open binary file."""
     with Path.open(RESOURCES / "IO_binary_file.bin", "rb") as open_binary_file:  # pylint: disable=unspecified-encoding
         open_binary_file.seek(56)
-        test_dict = {}
+        test_dict: dict[Any, Any] = {}
         byte_size = gwy.gwy_read_component(initial_byte_pos=56, open_file=open_binary_file, data_dict=test_dict)
         assert byte_size == 73
         assert list(test_dict.keys()) == ["test object component"]

--- a/tests/test_ibw.py
+++ b/tests/test_ibw.py
@@ -28,7 +28,7 @@ def test_load_ibw(
     result_pixel_to_nm_scaling = float
 
     file_path = RESOURCES / file_name
-    result_image, result_pixel_to_nm_scaling = load_ibw(file_path, channel)
+    result_image, result_pixel_to_nm_scaling = load_ibw(file_path, channel)  # type: ignore
 
     assert result_pixel_to_nm_scaling == pixel_to_nm_scaling
     assert isinstance(result_image, np.ndarray)

--- a/tests/test_io.py
+++ b/tests/test_io.py
@@ -8,6 +8,8 @@ import h5py
 
 from AFMReader.io import unpack_hdf5
 
+# mypy: disable-error-code="index"
+
 
 def test_unpack_hdf5_all_together_group_path_default(tmp_path: Path) -> None:
     """Test loading a nested dictionary with arrays from HDF5 format with group path as default."""

--- a/tests/test_jpk.py
+++ b/tests/test_jpk.py
@@ -32,7 +32,7 @@ def test_load_jpk(
     result_pixel_to_nm_scaling = float
 
     file_path = RESOURCES / file_name
-    result_image, result_pixel_to_nm_scaling = load_jpk(file_path, channel)
+    result_image, result_pixel_to_nm_scaling = load_jpk(file_path, channel)  # type: ignore
 
     assert result_pixel_to_nm_scaling == pixel_to_nm_scaling
     assert isinstance(result_image, np.ndarray)

--- a/tests/test_spm.py
+++ b/tests/test_spm.py
@@ -1,7 +1,7 @@
 """Test the loading of spm files."""
 
 from pathlib import Path
-from unittest.mock import patch
+from unittest.mock import patch, MagicMock
 
 import numpy as np
 import pySPM
@@ -58,7 +58,7 @@ def test_load_spm(
     ],
 )
 def test__spm_pixel_to_nm_scaling(
-    mock_pxs,
+    mock_pxs: "MagicMock",
     spm_channel_data: pySPM.SPM.SPM_image,
     filename: str,
     unit: str,

--- a/tests/test_topostats.py
+++ b/tests/test_topostats.py
@@ -44,6 +44,6 @@ def test_load_topostats(
     assert result_pixel_to_nm_scaling == pixel_to_nm_scaling
     assert isinstance(result_image, np.ndarray)
     assert result_image.shape == image_shape
-    assert set(result_data.keys()) == data_keys
+    assert set(result_data.keys()) == data_keys  # type: ignore
     assert result_data["topostats_file_version"] == topostats_file_version
     assert result_image.sum() == image_sum

